src/Client/JWTProvider.cpp

```diff
@@ -0,0 +1,280 @@
+    Poco::Net::HTTPRequest request(Poco::Net::HTTPRequest::HTTP_POST, token_url.getPathAndQuery(), Poco::Net::HTTPMessage::HTTP_1_1);
+    request.setContentType("application/x-www-form-urlencoded");
+
+    std::string request_body = "grant_type=refresh_token&client_id=" + oauth_client_id + "&refresh_token=" + idp_refresh_token;
+    request.setContentLength(request_body.length());
+    session->sendRequest(request) << request_body;
+
+    Poco::Net::HTTPResponse response;
+    std::istream & rs = session->receiveResponse(response);
+    if (response.getStatus() != Poco::Net::HTTPResponse::HTTP_OK)
+    {
+        std::string error_body;
+        Poco::StreamCopier::copyToString(rs, error_body);
+        idp_refresh_token.clear();
+        throw Exception(ErrorCodes::NETWORK_ERROR, "Error refreshing token: {} {}\nResponse: {}", response.getStatus(), response.getReason(), error_body);
+    }
+
+    Poco::JSON::Object::Ptr object = Poco::JSON::Parser().parse(rs).extract<Poco::JSON::Object::Ptr>();
+    idp_access_token = object->getValue<std::string>("access_token");
+    idp_access_token_expires_at = Poco::Timestamp::fromEpochTime(jwt::decode(idp_access_token).get_payload_claim("exp").as_integer());
+    if (object->has("refresh_token"))
+            idp_refresh_token = object->getValue<std::string>("refresh_token");
+}
+
+std::unique_ptr<Poco::Net::HTTPSClientSession> JWTProvider::createHTTPSession(const Poco::URI & uri)
+{
+    if (uri.getScheme() == "https")
+    {
+        Poco::Net::Context::Ptr context = Poco::Net::SSLManager::instance().defaultClientContext();
+        return std::make_unique<Poco::Net::HTTPSClientSession>(uri.getHost(), uri.getPort(), context);
+    }
+    throw Exception(ErrorCodes::SUPPORT_IS_DISABLED, "Built without SSL, ClickHouse cannot use JWT authentication without SSL support.");
+}
+
+void JWTProvider::openURLInBrowser(const std::string & url)
+{
+    if (url.empty())
+        return;
+
+#if defined(OS_DARWIN) || defined(OS_LINUX)
+    std::string command;
+#if defined(OS_DARWIN)
+    command = "open";
+#elif defined(OS_LINUX)
+    command = "xdg-open";
+#endif
+
+    if (command.empty())
+        return;
+
+    pid_t pid;
+    const char * argv[] = {command.c_str(), url.c_str(), nullptr};
+    int status = posix_spawnp(&pid, command.c_str(), nullptr, nullptr, const_cast<char * const *>(argv), nullptr);
+
+    if (status == 0)
+    {
+        int wait_status;
+        waitpid(pid, &wait_status, 0);
+    }
+#elif defined(OS_WINDOWS)
+    ShellExecuteA(NULL, "open", url.c_str(), NULL, NULL, SW_SHOWNORMAL);
+#endif
+}
+
+Poco::Timestamp JWTProvider::getJwtExpiry(const std::string & token)
+{
+    if (token.empty())
+        return 0;
+
+    try
+    {
+        auto decoded_token = jwt::decode(token);
+        return Poco::Timestamp::fromEpochTime(decoded_token.get_payload_claim("exp").as_integer());
+    }
+    catch (...)
+    {
+        return 0;
+    }
+}
+
+bool isCloudEndpoint(const std::string & host)
```

Comment:
There is a similar check here
https://github.com/ClickHouse/ClickHouse/blob/0cc644f9137389b37651ee8b7e602accb2dc68fd/src/Client/ConnectionParameters.cpp#L36

Maybe we can reuse this function and move it out of JWT